storybookjs/frontpage
# Docs SubNav and CodeSnippets improvements

> **Note for reviewers**: Strongly recommend reviewing commit-by-commit. I attempted to keep it as organized as I could for that purpose.

## What I did

![image](https://user-images.githubusercontent.com/486540/224155872-bc129e3a-9450-467c-bb64-4ec6f83e5ae1.png)

- Make docs SubNav sticky
- Store framework (renderer, actually, but it's called framework in the code) and code language in `localStorage`
- Make `/docs` redirect to `/docs/FRAMEWORK/get-started/install/`, where FRAMEWORK either comes from `localStorage` or falls back to the default, `react`
- Refactor CodeSnippets to filter by code language in addition to framework
    - And change default to TypeScript
- Add a code language selector that sets and reads from the `localStorage` value

And here's the how that breaks down per-commit:

- Make docs SubNav sticky
- Set selected framework in localStorage
- Remove Netlify redirect for `/docs`
    - Will handle client-side to respect framework in localStorage
- Add client-side `/docs` redirect
    - Respects framework found in localStorage
- Extract & test `getResolvedPaths` in CodeSnipppets
    - Add & configure `ts-jest`
    - Convert a couple troublesome files to `.ts`
- Add CodeLanguageSelector; refactor CodeSnippets
    - Add CodeLanguageSelector to SubNav
    - Store value in localStorage
    - CodeSnippets reads selected code language from localStorage and filters displayed tabs on it (in addition to current framework)
    - Refactored logic for resolving snippet paths is now tested
- Switch to more advanced `useLocalStorage`
    - This one listens for a change event, so it can be used as global state
- Only add TS 4.9 code language option in v7+
- Update `@storybook/components-marketing`
    - To get this fix: https://github.com/storybookjs/components-marketing/pull/62
- Update circleci node image
    - The `circleci/node:16` image only went up to `node@16.13`
    - One of our dependencies, `usehooks-ts`, requires `node@^16.15`
        - [See build error](https://app.circleci.com/pipelines/github/storybookjs/frontpage/2880/workflows/de83ce03-f8cc-4c91-b08e-25702e26afc8/jobs/2792?invite=true#step-103-31)
    - `circleci` image namespace is now legacy
        - https://discuss.circleci.com/t/legacy-convenience-image-deprecation/41034
    - `cimg` is the next-gen namespace
        - https://circleci.com/developer/images/image/cimg/node#image-tags

## How to test

_As these changes involve `localStorage`, this PR should receive some **rigorous** manual QA:_

1. Navigate to the [deploy preview homepage](https://deploy-preview-497--storybook-frontpage.netlify.app/)
    - Or spin it up locally with `GATSBY_SKIP_ADDON_PAGES=true yarn start`
2. Open DevTools
    1. Open the Application tab at the top
    2. Then open the Local Storage menu under Storage in the right-hand sidebar
    3. Select the current domain
    ![Screen Shot 2023-03-09 at 11 36 08 AM](https://user-images.githubusercontent.com/486540/224124380-53ebdfce-df52-43d8-9d4b-e015d251f43f.png)

3. **Confirm** that the `sb-docs-selected-framework` & `sb-docs-selected-code-language` keys are not present in Local Storage
    - If they _are_ present, please right-click on each of them and choose Delete from the context menu to start this process fresh
4. Click on either Docs > Guides or the Get started button, both which link to `/docs`
5. **Confirm** that you are redirected to `/docs/react/get-started/install/`
    - Because React is the default framework
6. **Confirm** that there are still no values for those keys in Local Storage
7. Navigate to the _What's a story?_ page
	- So that we have heading anchors to test
8. Scroll down until the _Edit a story_ heading is just visible at the bottom of your viewport
   ![image](https://user-images.githubusercontent.com/486540/224128280-82f4e124-c33f-4a47-a052-c42fbeb47bd6.png)
10. **Confirm** that the SubNav bar is now sticky to the top of the viewport
   ![image](https://user-images.githubusercontent.com/486540/224128518-37722738-f9aa-47b1-94c0-f95013c42782.png)
12. Click the _Edit a story_ heading's anchor link
13. **Confirm** that it jumps to the top of the page while leaving room for the SubNav bar
   ![image](https://user-images.githubusercontent.com/486540/224128096-239ab20a-b72b-4e35-a286-e59808908244.png)
14. Use the framework selector (it'll be labeled "React" at this point) to choose the Angular framework
15. **Confirm** that the code language selector disappears and the code snippets use TS
14. Use the framework selector (it'll be labeled "React" at this point) to choose any framework besides React or Angular
    - In these instructions, I'll use FRAMEWORK
16. **Confirm** that you are redirected to FRAMEWORK's version of that page
17. **Confirm** that the both the framework selector and the code snippets reflect that framework value
18. **Confirm** that the `sb-docs-selected-framework` key is set to `"FRAMEWORK"` in Local Storage
   ![image](https://user-images.githubusercontent.com/486540/224127813-0dcece32-71f8-4bb8-8fd9-8775f322a5ff.png)
19. Click the Storybook logo to navigate to the homepage
20. Click on either Docs > Guides or the Get started button, both which link to `/docs`
21. **Confirm** that you are redirected to `/docs/FRAMEWORK/get-started/install/`
    - Because now FRAMEWORK is your chosen framework
22. Click the Storybook logo to navigate to the homepage
23. Right-click on the value of the `sb-docs-selected-framework` key in Local Storage and select Edit "Value" in the context menu
24. Change it to an invalid framework value, e.g. `"invalid"` (the quotes are required)
25. Click on either Docs > Guides or the Get started button, both which link to `/docs`
26. **Confirm** that you are redirected to `/docs/react/get-started/install/`
    - Because React is the default framework
27. **Confirm** that the `sb-docs-selected-framework` key is set to `"react"` in Local Storage
    - Because React is the default framework, to which we coerce bad values
28. Navigate to the _Write stories > Introduction_ page
    - So we can see lots of snippets
29. **Confirm** that all of the snippets show the TS version
    - Some snippets for some frameworks may not have a TS version, in which case they will show the JS version and message explaining why that prompts a contribution
30. Use the language selector (it'll be labeled "TS" at this point) to choose JavaScript
31. **Confirm** that TS 4.9 is not an available option
    - This will enable for `7.0` docs, which aren't available in a deploy preview
32. **Confirm** that the snippets update from TS to JS
33. **Confirm** that the `sb-docs-selected-code-language` key is set to `"js"` in Local Storage
34. Navigate to another page
35. **Confirm** that the code language selector is still set to JS and the code snippets render the JS version
36. Click the Storybook logo to navigate to the homepage
37. Right-click on the value of the `sb-docs-selected-code-language` key in Local Storage and select Edit "Value" in the context menu
38. Change it to an invalid code language value, e.g. `"invalid"` (the quotes are required)
39. Click on either Docs > Guides or the Get started button, both which link to `/docs`
40. **Confirm** that the `sb-docs-selected-code-language` key is set to `"ts"` in Local Storage
    - Because TS is the default code language, to which we coerce bad values
41. **Confirm** that both the code language selector and the code snippets reflect that TS value

_Anything else you can think of to test._ Especially:
- Navigating around to various pages for various frameworks, looking for broken/missing/unexpected snippets
